core/src/test/scala/org/apache/spark/sql/delta/DeltaSuite.scala

```diff
@@ -596,6 +596,95 @@ class DeltaSuite extends QueryTest
     }
   }
 
+  test("batch write: append, dynamic partition overwrite") {
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq(1, 2, 3).toDF
+        .withColumn("part", $"value" % 2)
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq(1, 5).toDF
+        .withColumn("part", $"value" % 2)
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("overwrite")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.toDF.select($"value".as[Int]), 1, 2, 5)
+
+      Seq(("a", "x"), ("b", "y"), ("c", "x")).toDF("value", "part")
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("overwrite")
+        .option("overwriteSchema", true)
+        .save(tempDir.getCanonicalPath)
+
+      Seq(("a", "x"), ("d", "x")).toDF("value", "part")
+        .write
+        .format("delta")
+        .partitionBy("part")
+        .mode("overwrite")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.toDF.select($"value".as[String]), "a", "b", "d")
+    }
+  }
+
+  test("batch write: append, dynamic partition overwrite with replaceWhere") {
+    withTempDir { tempDir =>
+      def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
+
+      Seq((1, "x"), (2, "y"), (3, "z")).toDF("value", "sub")
+        .withColumn("part", $"value" % 2)
+        .write
+        .format("delta")
+        .partitionBy("part", "sub")
+        .mode("append")
+        .save(tempDir.getCanonicalPath)
+
+      Seq((3, "x"), (5, "x")).toDF("value", "sub")
+        .withColumn("part", $"value" % 2)
+        .write
+        .format("delta")
+        .partitionBy("part", "sub")
+        .mode("overwrite")
+        .option(DeltaOptions.REPLACE_WHERE_OPTION, "part = 1")
+        .option("partitionOverwriteMode", "dynamic")
+        .save(tempDir.getCanonicalPath)
+      checkDatasetUnorderly(data.toDF.select($"value".as[Int], $"sub".as[String]),
+        (2, "y"), (3, "x"), (5, "x"))
+    }
+  }
+
+  test("batch write: append, overwrite without partitions should ignore partition overwrite mode") {
```

Comment:
since partitionOverwriteMode can also be set in sparkconf generally i am hesitant to throw error or log warning. the user might simply have set it for other partitioned writes and not meant to target this one at all.

now if it was set in the conf for this particular write then i agree maybe an error or warning is appropriate. but this requires us to start teasing apart what was set in conf and in spark conf (which can be done).

Reply:
> since partitionOverwriteMode can also be set in sparkconf generally i am hesitant to throw error or log warning. the user might simply have set it for other partitioned writes and not meant to target this one at all.

I agree. We can think about possibly logging a warning if it's set for the particular write.